src/khoj/processor/url/web_crawler.py

```diff
@@ -0,0 +1,58 @@
+# The following code is heavily inspired by the implementation found at: [Auto-GPT](
+# https://github.com/Significant-Gravitas/Auto-GPT/blob/master/autogpt/commands/web_selenium.py)
+
+from bs4 import BeautifulSoup
+from selenium.common.exceptions import WebDriverException
+
+# from selenium.webdriver.chrome.options import Options as ChromeOptions
+# from selenium.webdriver.chrome.service import Service as ChromeDriverService
+# from selenium.webdriver.chrome.webdriver import WebDriver as ChromeDriver
+from selenium.webdriver.common.by import By
+from selenium.webdriver.common.options import ArgOptions as BrowserOptions
+from selenium.webdriver.edge.options import Options as EdgeOptions
+from selenium.webdriver.edge.service import Service as EdgeDriverService
+from selenium.webdriver.edge.webdriver import WebDriver as EdgeDriver
+
+# from selenium.webdriver.firefox.options import Options as FirefoxOptions
+# from selenium.webdriver.firefox.service import Service as GeckoDriverService
+# from selenium.webdriver.firefox.webdriver import WebDriver as FirefoxDriver
+from selenium.webdriver.remote.webdriver import WebDriver
+
+# from selenium.webdriver.safari.options import Options as SafariOptions
+# from selenium.webdriver.safari.webdriver import WebDriver as SafariDriver
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.support.wait import WebDriverWait
+
+# from webdriver_manager.chrome import ChromeDriverManager
+# from webdriver_manager.firefox import GeckoDriverManager
+from webdriver_manager.microsoft import EdgeChromiumDriverManager as EdgeDriverManager
+
+
+def get_webdriver() -> WebDriver:
+    options: BrowserOptions = EdgeOptions()
+    options.add_argument(
+        "user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.5615.49 Safari/537.36"
+    )
+    options.add_argument("--headless")
+    options.add_experimental_option("excludeSwitches", ["enable-logging"])
+    driver = EdgeDriver(service=EdgeDriverService(EdgeDriverManager().install()), options=options)
+    return driver
+
+
+def scrape_text_with_selenium(driver: WebDriver, url: str) -> str:
+    driver.get(url)
+
+    WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.TAG_NAME, "body")))
+
+    # Get the HTML content directly from the browser's DOM
+    page_source = driver.execute_script("return document.body.outerHTML;")
```

Comment:
It would be better to use a library like `langchain` to hold some of this scraping logic and data extraction. See here: https://python.langchain.com/docs/integrations/document_loaders/web_base.

Web scrapers are notoriously brittle and hard to negotiate over time. If we can use a dependency that handles this more reliably, we can spend less time managing this integration.